src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectMethodExecutor.cs

```diff
@@ -132,8 +149,16 @@ private static ActionExecutor WrapVoidAction(VoidActionExecutor executor)
             };
         }
 
-        private static ActionExecutorAsync GetExecutorAsync(MethodInfo methodInfo, TypeInfo targetTypeInfo)
+        private static ActionExecutorAsync GetExecutorAsync(Type taskInnerType, MethodInfo methodInfo, TypeInfo targetTypeInfo)
         {
+            if (taskInnerType == null)
+            {
+                // This will be the case for types which have derived from Task and Task<T> or non Task types.
+                throw new InvalidOperationException(Resources.FormatActionExecutor_UnexpectedTaskInstance(
+                    methodInfo.Name,
+                    methodInfo.DeclaringType));
+            }
```

Comment:
Why do we need to throw an exception in this private method - what code would a user write to hit this?